File: leetcode2/second/0328. Odd Even Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def oddEvenList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        oddnode,evennode=ListNode(0),ListNode(0)
        podd,peven=oddnode,evennode
        while head and head.next:
            podd.next=head
            peven.next=head.next
            podd=podd.next
            peven=peven.next
            head=head.next.next
        if head:
            podd.next=head
            podd=podd.next
        peven.next=None
        podd.next=evennode.next
        return oddnode.next

File: archives/C#/0328. Odd Even Linked List.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode OddEvenList(ListNode head) {
        if(head==null || head.next==null){
            return head;
        }
        ListNode pOdd=head;
        ListNode pEvenPart=head.next;
        ListNode pEven=pEvenPart;
        while (pEven.next!=null && pEven.next.next!=null){
            pOdd.next=pEven.next;
            pOdd=pOdd.next;
            pEven.next=pOdd.next;
            pEven=pEven.next;
        }
        if(pEven.next!=null){
            pOdd.next=pEven.next;
            pOdd=pOdd.next;
        }
        pEven.next=null;
        pOdd.next=pEvenPart;
        return head;
    }
}
